openverse-api/catalog/api/utils/waveform.py

```diff
@@ -0,0 +1,131 @@
+import json
+import logging
+import math
+import mimetypes
+import os
+import pathlib
+import shutil
+import subprocess
+
+import requests
+
+log = logging.getLogger(__name__)
+
+TMP_DIR = pathlib.Path('/tmp').resolve()
+UA_STRING = 'OpenverseWaveform/0.0 (https://wordpress.org/openverse)'
+
+
+def ext_from_url(url):
+    """
+    Get the file extension from the given URL. Looks at the last part of the URL
+    path, and returns the string after the last dot.
+
+    :param url: the URL to the file whose extension is being determined
+    :returns: the file extension or ``None``
+    """
+    file_name = url.split('/')[-1]
+    if '.' in file_name:
+        ext = file_name.split('.')[-1]
+        return f'.{ext}'
+    else:
+        return None
+
+
+def download_audio(url, identifier):
+    """
+    Download the audio from the given URL to a location on the disk.
+
+    :param url: the URL to the file being downloaded
+    :param identifier: the identifier of the media object to name the file
+    :returns: the name of the file on the disk
+    """
+    log.info(f'Downloading file at {url}')
+
+    headers = {
+        'User-Agent': UA_STRING
+    }
+    with requests.get(url, stream=True, headers=headers) as res:
+        log.debug(f'Response code: {res.status_code}')
+        mimetype = res.headers['content-type']
+        log.debug(f'MIME type: {mimetype}')
+        ext = ext_from_url(url) or mimetypes.guess_extension(mimetype)
+        if ext is None:
+            raise ValueError('Could not identify media extension')
+        file_name = f'audio-{identifier}{ext}'
+        log.debug(f'File name: {file_name}')
+        with open(TMP_DIR.joinpath(file_name), 'wb') as file:
+            shutil.copyfileobj(res.raw, file)
```

Comment:
Is there any concern that requests to this endpoint could balloon the temp dir?

Reply:
Yes, it is a theoretical possibility. But if everything goes smoothly, the `cleanup()` function deletes the downloaded file after the waveform is generated.